### ISSUE TITLE
pronouns on profile as first class citizens. 

### ISSUE THREAD

**Issue by shleeable** [tootsuite/mastodon#16226]
### Pitch
<img width="589" alt="image" src="https://user-images.githubusercontent.com/17537000/117953943-20404700-b349-11eb-9fb0-559dd5a86422.png">

### Motivation
This is the way it's going to go, and we might as well get out front of history.

**Comment by shleeable** [comment]
lol. Apparently I asked for this in #12261 in 2019.

**Comment by SuperSandro2000** [comment]
Why can't Profile metadata or bio be used for this?

**Comment by codl** [comment]
If I remember right, pronouns were part of the push to get arbitrary fields on profiles in the first place. As far as my slice of the fediverse is concerned, metadata fields have done the job fine for years. **I think this is worth considering**, though.

"This is the way it's going to go" is a bad justification for this, or any, feature, so let me offer another: 

Having an explicit pronouns field in the profile editor normalizes putting pronouns on one's profile, having the implied option of doing it with fields or in one's bio does not. Right now, putting pronouns in your bio is solely encouraged by some communities. By making this change, it could be encouraged by the software itself.

Matters are complicated for people for whom the proposed 17-character field could be too short:
* genderfluid people and others who use different pronouns at different times or even at once
* plural systems whose different members use different pronouns
* multilingual people who may want to document their pronouns in many languages, along with grammatical gender
* more?

I've seen people use up to all four metadata fields for those kinds of complicated pronoun sets, so just making the field longer wouldn't work. Of course those people would likely keep using the fields, but I'm worried about new users in that situation being directed to use a field that works for most but not for them.

There's another thing I'm worried about, a problem that metadata fields had too in their time: clients and other AP implementations will take time to add support, and until then you can't really rely on the feature to actually deliver the information you put into it. I remember a time when you couldn't really put important things in your metadata (like your pronouns!) because a majority of mobile clients simply didn't display metadata. Perhaps one way to alleviate this is to, in the backend, coopt a metadata field and call it "Pronouns", and make a special case in the front-end to display it differently, so any client or AP implementation that already supports metadata will at least display it.

Further reading:
* #232, the initial request for what became metadata fields. OP makes the same point there that I did here re normalizing pronouns.
* #3211, a much more elaborate request for a pronoun field, which was closed when metadata fields came out. Much discussion of "complicated" pronoun sets there

And, re scraping: this is already very easy to do. Everyone writes pronouns in the same, conveniently regexable way.

**Comment by codl** [comment]
A friend pointed out a potential way to solve the *field too short/field needs more lines* problem: have it be like a git commit message

Simple multiline input (or potentially a single-line "main" input and a multiline "more" input)
First line is limited in length (17 characters?) and is the part that is displayed on the profile or wherever.
An indicator appears when more than one line exists, and tapping/clicking it reveals the rest of the field.

I am not sure this could work in the webapp, though. I can't think of anything that expands in place or opens a modal with just a few lines of text in the mastodon UI. Also this is a lot of complexity for something that people are already solving with metadata fields / their bio.

**Comment by Brawaru** [comment]
How do pronouns work with other languages than English? /gen

In Russian we only have three pronouns, one of which is neuter — it, and you really don't want to apply _it_ to people, because it would be offensive, _it_'s not even often applied to animals whose gender is unknown here. I am not aware of any other pronouns in Russian (created or adopted). Singular ‘they’ would be considered here a mistake, it sounds very off too (thanks to declension, probably), even though there were attempts to adopt it. 

This, at least, is already an issue, meaning pronouns cannot be shared across languages, yet Mastodon is an international software. This is why I personally find that manually putting pronouns in bio or meta field is more preferable solution, because it's untranslatable content.

**Comment by Cassolotl** [comment]
>How do pronouns work with other languages than English? /gen

@Brawaru [This Wikipedia page) might help?](https://en.wikipedia.org/wiki/English_personal_pronouns)

>Matters are complicated for people for whom the proposed 17-character field could be too short:

I feel like that might be *everyone*? :D In English, each pronoun has five forms. A neopronoun set, which others might not be familiar with, would need to display all five forms in order to be useable. So, for singular they it'd be "they/them/their/theirs/themself" - that's 31 characters, I think? He/him/his/his/himself is 22. Elverson, e/em/eir/eirs/emself, is 20. I don't know of any common sets shorter than that!

**Comment by codl** [comment]
> This is why I personally find that manually putting pronouns in bio or meta field is more preferable solution, because it's untranslatable content.

I don't believe OP was proposing a list of pronouns which would be translated to every language, but rather a free entry text field dedicated to pronouns. I agree with you that attempting to translate all pronouns in all languages would be silly.

>> people for whom the proposed 17-character field could be too short
>
> I feel like that might be _everyone_?

How is it a problem for everyone? I see a lot of people, including people with neopronouns, only display the first two forms or sometimes two first forms i.e. "she/fae", and they aren't hurting for space to add the rest. If they had a dedicated pronoun field, they'd write the same thing into it regardless of how much space they had.

Either way I don't know why OP went with 17 characters, I just went along because it seems like it'd be easier to fit in UI. I guess it all depends on where exactly this field would show up. OP mocked up in #12261 a version that showed a badge on each post, which I think is neat but would absolutely need a length limit. I think the "commit message" behaviour I described above could work well with this, with the badge being a "she/fae" style reminder that could be expanded to see the whole declension or whatever other details the user has added. But if this only shows up on someone's profile like an extra line of metadata then, who cares, give everyone 10k characters.

**Comment by shleeable** [comment]
Pixelfed doing the right thing. https://github.com/pixelfed/pixelfed/commit/fabb57a9

**Comment by ryliejamesthomas** [comment]
> Pixelfed doing the right thing. [pixelfed/pixelfed@fabb57a](https://github.com/pixelfed/pixelfed/commit/fabb57a9)

I really don't think limiting pronouns to a predefined set is the right solution. That seems to be what Pixelfed's doing; or am I misunderstanding?

**Comment by ryliejamesthomas** [comment]
I do agree a dedicated pronouns section could be useful though. For example you could show a person's pronouns if you hover over their avatar in the timeline, or display a pronoun reminder somewhere when you are replying to someone.

**Comment by shleeable** [comment]
<img width="466" alt="image" src="https://user-images.githubusercontent.com/17537000/121844625-d4e9d180-cd16-11eb-86d7-e6da4a7a977a.png">

Slack doing the right tihng.

**Comment by shleeable** [comment]
> > Pixelfed doing the right thing. [pixelfed/pixelfed@fabb57a](https://github.com/pixelfed/pixelfed/commit/fabb57a9)
> 
> I really don't think limiting pronouns to a predefined set is the right solution. That seems to be what Pixelfed's doing; or am I misunderstanding?

They are using a predefinited set, but it's a step in the right direction. So he gets a cookie.

**Comment by shleeable** [comment]
<img width="510" alt="image" src="https://user-images.githubusercontent.com/17537000/122663041-39e77080-d1ca-11eb-9ff7-9c95bcf79534.png">

Linkedin doing the right thing

**Comment by Wuzzy2** [comment]
I like this idea. Not because I personally need this feature, but because I know that pronouns are kind of a big deal for parts of trans, non-binary, genderfluid, and related communities. To some, being called by the right pronoun, including online, is very important for the emotional well-being. It’s not just a cringe woke "Tumblr trend", that's just a myth. Deliberate misgendering is violence. There’s solid science that backs the importance of this topic up, for example: https://www.jahonline.org/article/S1054-139X(18)30335-5/fulltext

I also think the pronoun field should be freetext, rather than a dropdown list, it is NOT realistically possible to cover all options with a drop-down list. Especially when you start to consider multiple languages, the list will get out of hand fast. A drop-down list also risks of baking a certain kind of language into the system, so it cannot react fast to changes to language (imagine when a neopronoun actually gains popularity and turns mainstream). Thus, freeform text is the simplest and most flexible option.

Also, it should be completely optional, many people don’t want to to disclose this, or don’t want to think about this, etc. Or they don’t care. This includes some cis people. This is all valid, too. If no pronouns are specified, then none are shown. Simple as that. So this debunks the argument that this feature will "force" people to add their pronouns, or that it’ll somehow lead to "othering".

I think a case for pronouns being used as a core feature, rather than using metadata for it, is, it's kinda related to one's name. You need the name to identify someone. And you need pronouns to refer someone. Since pronouns are so intrinsically linked to someone's identity, it should be a core feature. Metadata, on the other hand, are more random. They can be anything, so it's not exactly comparable. Pronouns can also be rendered in a different way, like next to the name or something like that.

I'm neutral on the *exact* implementation of this feature, however. Like how long this field should be. 17 is a very arbitrary number. Could also be any other number in theory.

**Comment by ClearlyClaire** [comment]
I thought I had already commented on this issue, but apparently not.

I think this is *mostly* addressed by profile fields, but I can see the benefit in having pronouns readily available without going to the person's profile. I can also see the benefit in encouraging people to explicitly select a pronoun.

However, pronouns and grammatical gender are heavily language-dependent, and even something as simple as a choice between “he”, “she”, “they” and “custom”, without any built-in neopronoun, may prove especially difficult to translate. Indeed, not all languages Mastodon supports or will eventually support might have gendered pronouns, and even amongst those which do, there might not be a consensus on a gender-neutral pronoun. For instance, a French has no neutral pronoun as widely accepted as the English “them” but multiple proposals such as « iel », « ael », « ul » or « ol ». A French user might want to be referred to as “them” in English and « iel » in French while another one might prefer being referred to as “them” in English and « ael » in French.

So, while presets would be handy and also allow for guidance on grammatical gender, I do not think they would work well given the multilingual nature of Mastodon.

A free-form text might work better, but then it would obviously not get translated.

As for compatibility with clients and other fediverse software, I am not too worried, as that could just be serialized the same way as current profile fields, but with extra attributes. This way, existing fediverse software would just show it as a generic profile field.

**Comment by ClearlyClaire** [comment]
The more I think of it, the more I think this issue comes down to two things:
- encouraging people speaking gendered language to put down how they'd like to be referred to in their profile: this could be achieved simply by changing the placeholders for the first profile metadata field; this also has the advantage of not affecting potential languages where that does not make sense (by using other placeholders in the translation)
- easily know how someone would like to be referred to when replying to them: I do not have a solution to this, but some observations:
  * it only matters for people you do not already interact with frequently
  * it mainly matters when replying to them
  * other things described in the profile might matter (e.g. request to not use a specific word, not talk about a specific topic, etc.)

**Comment by ClearlyClaire** [comment]
Random design ideas based on those thoughts:

- warning showing up when replying to someone with no follow relationship:
![image](https://user-images.githubusercontent.com/384364/146784732-beb6e8cb-5a2a-4867-877d-1d5562314326.png)

- settings page:
![image](https://user-images.githubusercontent.com/384364/146784997-c3ecfc60-5de7-420a-b7d2-04bc20e86333.png)

**Comment by ClearlyClaire** [comment]
I think https://github.com/mastodon/mastodon/pull/17151 and https://github.com/mastodon/mastodon/pull/17152 may address this issue.

**Comment by ryliejamesthomas** [comment]
Hmmm. I think this is a good step, but still think it's much more useful to show the information rather than tell someone to go find it.

If you don't want a specific pronouns field, perhaps you could have a toggle/popup/whatever to show the whole metadata table (or just show it)? It's not like the left column is lacking space (can't remember what replying looks like in single-column mode, though). But I get that pointing to the profile could be more desirable than bunging in the whole bio, which could be quite long.

I also think that, as a person's pronouns can change, familiarity isn't necessarily the best trigger for the prompt.

Does the link to the person's profile open in a new column, or replace the one with the conversation the person's replying to?

**Comment by OneJaeAtATime** [comment]
I would like to add that this would be a great idea that would go a long way to helping normalise pronoun usage.

**Comment by Cassolotl** [comment]
I've recently noticed a few high-profile TERFs making their way over to Mastodon as Twitter implodes. They love to bang on about how "pronouns" reinforce [something something blabla transphobic bigotry], and sometimes say things like "I'm going to boycott [company] for having pronoun badges for employees" or "I'm refusing to put my pronouns in my email signature at work".

So, if we have a pronoun field on profiles, I bet TERFs will be less likely to migrate to Mastodon. >:D

**Comment by afontenot** [comment]
I like the idea of continuing to use profile metadata for pronouns, because there's already an understood mechanism for syncing this information cross-instance. On the other hand, there should probably be some special treatment for the pronoun field.

For example, it could be a fifth "free" metadata field with an uneditable name. (The name would change depending on the UI translation, but it wouldn't be user editable.) This would make it unambiguous to Mastodon code when a user has entered pronouns. The advantage of this is that Mastodon could then display pronouns in the UI at vital points, for instance (a) above the message box when you are replying to someone for the first time, and (b) in the preview of the user's profile when hovering over their username if https://github.com/mastodon/mastodon/issues/14918 was to be implemented.

Obviously, care would need to be taken to make sure this behaves correctly in all languages. On the other hand, having a fixed field might still be appropriate because quite a few users communicate on Mastodon in multiple languages. Just because a user's UI language doesn't use gendered pronouns doesn't mean that other languages they speak are similar.

**Comment by ThatOneCalculator** [comment]
The fifth free idea wouldn't work with federation, given services like Misskey allow people to add up to 16 metadata slots.

**Comment by afontenot** [comment]
> The fifth free idea wouldn't work with federation, given services like Misskey allow people to add up to 16 metadata slots.

I think the underlying protocol extension allows unlimited. By "free" I just mean that Mastodon would, by default, start allowing 5 instead of 4. One of these slots would only allow pronouns.

**Comment by 12345axc** [comment]
I think we should take account that some language doesn't have gendered pronoun :))
Like Indonesian who only have "dia" for every gender.

Some pronoun concept also not easily translateble into English language.

**Comment by smiba** [comment]
I personally think the profile metadata fields are fine for this and make sense. Adding a separate pronoun feature would add clutter and would personally not feel much different then what we have right now? 

Making the pronouns more prominent (more then the metadata fields currently are) would not result in more people gendering you correctly, the assholes will still misgender people on purpose and the people that care just the bare minimum will do just fine with the metadata field. 

Not sure if it matters, but as a trans person myself I think the current state of how you can add pronouns to mastodon is fine. 
Either that, or add more predefined fields to make this a more global change. For example you could make a bunch of options for the metadata fields like "pronouns", "location", "website" etc. where people can select from.

**Comment by violet-athena** [comment]
> Making the pronouns more prominent (more then the metadata fields currently are) would not result in more people gendering you correctly, the assholes will still misgender people on purpose and the people that care just the bare minimum will do just fine with the metadata field.

I would like to point out that including pronouns is not in order to gender people correctly but to make it a social convention therefore making the lives of people that need to use them for gendering easier and less embarrassing.

TikTok implemented this feature recently and it works great. It's neither intrusive, nor cluttered, nor anything like it.

![IMG_0352](https://user-images.githubusercontent.com/1032515/212074850-4cf13362-3e2a-4c8a-b19b-0a33d6ecccef.jpeg)

I can see an argument for keeping it in the metadata, but considering how ugly things around this are becoming, showing support in this way is not a big ask.

**Comment by smiba** [comment]
TikTok however does not force you to enter this information, and when the information is not entered (or if you're in a country this is not enabled for, for example in The Netherlands I have no pronoun option) it's simply blank.

This is not really much different then from what you can already do on mastodon right now, most instances (or at least the LGBT friendly ones) will have pronoun emojis you can put in your name. 

I should note that I'm not outright against this proposal, but I do believe that if a change is made for pronouns it should be considered carefully as not to take up UI space (especially if a user keeps it blank, it should not take up space) or ends up being too invasive to the user experience. 

EDIT: I no longer feel as strong about it as I felt when I wrote this, I think the pronoun field might be beneficial especially if there is multilingual support

**Comment by IRod22** [comment]
Here is what I'm thinking about this conversation. As @smiba [said](https://github.com/mastodon/mastodon/issues/16226#issuecomment-1376532121), there will still be users that disrespect a user's pronouns unless the victim reports the disrespectful user(s), and admins on either side take action against those disrespectful user(s). I still think that pronouns should be implemented because those are becoming the norm, and I think I have an idea on how they could be implemented on user profiles and posts.

On a user profile or post, the pronouns could be right next to the user's Mastodon handle like this:
```jsx
<span><a href={`/@${user.username}@${user.domain}`}>@{user.username}@{user.domain}</a>{user.pronouns && <> -- {user.pronouns}</>}</span>
```

On the server, a `pronouns` model property could be implemented as a text-based field on the `Account` model.

**Comment by jepyang** [comment]
One argument in favor of a dedicated pronoun field that I haven’t seen mentioned yet: accessibility.

With the current set up, there are lots of people listing pronouns but little consensus about where to put that info. Some people use the metadata fields, some people include it in their display name, some people have it in their bio, and occasionally it’ll even end up in a pinned post and nowhere else. Sometimes it’s plaintext, sometimes it’s a custom emoji. Sometimes it has a label, sometimes it’s just the pronouns by themselves. Etc.

I can only speak for myself, but my particular flavor of ADHD/neurodiversity can make it surprisingly difficult to pull someone’s pronoun info out of the pile of info their profile presents me with. I’d guess that other forms of disability might also interfere (visual processing disorders, dyslexia, maybe vision impairment/screen reader use?) with locating this information.

I don’t know how anyone else relates to pronouns but to me it feels like crucial information for interacting with people online, second only to their name. Having them visible in a consistent place (and by extension, making it less ambiguous when someone hasn’t listed any pronouns, instead of being left wondering if their pronouns are listed but just buried in their profile/bio) would be a massive improvement in usability from my perspective.

Doesn't have to be mandatory, doesn’t have to be translated, doesn’t have to contain a drop-down list of all possible pronouns. Just a text field, explicitly for pronouns, that always displays in the same place, not unlike the way everyone has a display name and an account handle and a profile pic and four arbitrary metadata fields that all display in the same place.

**Comment by derspyy** [comment]
> One argument in favor of a dedicated pronoun field that I haven’t seen mentioned yet: accessibility.
> 
> With the current set up, there are lots of people listing pronouns but little consensus about where to put that info. Some people use the metadata fields, some people include it in their display name, some people have it in their bio, and occasionally it’ll even end up in a pinned post and nowhere else. Sometimes it’s plaintext, sometimes it’s a custom emoji. Sometimes it has a label, sometimes it’s just the pronouns by themselves. Etc.
> 
> I can only speak for myself, but my particular flavor of ADHD/neurodiversity can make it surprisingly difficult to pull someone’s pronoun info out of the pile of info their profile presents me with. I’d guess that other forms of disability might also interfere (visual processing disorders, dyslexia, maybe vision impairment/screen reader use?) with locating this information.
> 
> I don’t know how anyone else relates to pronouns but to me it feels like crucial information for interacting with people online, second only to their name. Having them visible in a consistent place (and by extension, making it less ambiguous when someone hasn’t listed any pronouns, instead of being left wondering if their pronouns are listed but just buried in their profile/bio) would be a massive improvement in usability from my perspective.
> 
> Doesn't have to be mandatory, doesn’t have to be translated, doesn’t have to contain a drop-down list of all possible pronouns. Just a text field, explicitly for pronouns, that always displays in the same place, not unlike the way everyone has a display name and an account handle and a profile pic and four arbitrary metadata fields that all display in the same place.

this is the closest to ideal as it can get. an optional field for pronouns.

a mandatory fields excludes some languages, as has been pointed out, and people who simply don't feel comfortable.
translated pronouns require a predefined list and almost no neo-pronouns have counterparts in other languages.
a drop-down list is just excessive in terms of accessibility, we should aim for simplicity.

though i have to point out that, for the time being, we as a community should push towards possibly using the first of the optional fields to write down pronouns, so it's the first thing to be parsed and possibly read by a screen reader. (i'm not very familiar)

additional rant:
i was going to suggest that third party clients should experiment with reading that field and displaying its value next to the user's name, but it'd have to be sure it's actually reading pronouns and parsing that would be hard.

**Comment by RealJosephKnapp** [comment]
Perhaps, although consider the effect of parallel economies. Those who do not think that neopronouns are a legitimate linguistic construct will remove them, thus creating two incompatible implementations of activitypub. The underlying protocol itself must be apolitical. Neopronouns are a politically charged wedge issue that the majority of the people in the world disagree with or have never heard of. If you conversed using neopronouns on masaii tribal lands, then masaii people would look at you funny. Use them in China and the native Chinese would call you a baizou. Use them in Russia and the Russians will laugh. You might consider neopronouns well and normal, but I don't and my bleeding heart liberal California extended family considers them to be restrictive and awkward to use. Know that your values are not universal and that the better solution is to increase the character limit. I say 5000 for toots and 7000 for profiles. That would leave plenty of space for as many pronouns as you could possibly need without risking fracturing the activitypub protocol on contentious wedge issues. It might also be desirable to add a link button that could take a word or sentence and turn it into a link that for example could redirect to HTTPS link, or a .onion link, or a ftp link, or a magnet link depending on what the link is copied and pasted into.

**Comment by ThatOneCalculator** [comment]
@derspyy I think it's less of an issue of practical implementation at this point and more of a technical implementation in regards to federation. Either adopting a *key-esque vCard field for it or a new ActivityPub for user profiles  (`toot:pronouns`, maybe?) would be best imo

**Comment by smiba** [comment]
I think I've said it before in this issue, but why not just use the current metadata fields? Why reinvent the wheel? 

Surely if you really wanted this to be a bit more of a prominent option have a selector for some pre-filled options to put in the metadata fields on your profile, including pronouns for example.

**Comment by derspyy** [comment]
> Perhaps, although consider the effect of parallel economies. Those who do not think that neopronouns are a legitimate linguistic construct will remove them, thus creating two incompatible implementations of activitypub. The underlying protocol itself must be apolitical. Neopronouns are a politically charged wedge issue that the majority of the people in the world disagree with or have never heard of. If you conversed using neopronouns on masaii tribal lands, then masaii people would look at you funny. Use them in China and the native Chinese would call you a baizou. Use them in Russia and the Russians will laugh. You might consider neopronouns well and normal, but I don't and my bleeding heart liberal California extended family considers them to be restrictive and awkward to use. Know that your values are not universal and that the better solution is to increase the character limit. I say 5000 for toots and 7000 for profiles. That would leave plenty of space for as many pronouns as you could possibly need without risking fracturing the activitypub protocol on contentious wedge issues. It might also be desirable to add a link button that could take a word or sentence and turn it into a link that for example could redirect to HTTPS link, or a .onion link, or a ftp link, or a magnet link depending on what the link is copied and pasted into.

this is not a discussion about neo-pronouns. regular everyday pronouns still exist and are part of everyday conversation in multiple languages, and having a way to check someone's pronouns will make it easier in case you are unsure on how someone is addressed. it seems to me that you are being narrow-minded, at best.

**Comment by derspyy** [comment]
> I think I've said it before in this issue, but why not just use the current metadata fields? Why reinvent the wheel?
> 
> Surely if you really wanted this to be a bit more of a prominent option, have a selector for some pre-filled options to put in the metadata fields on your profile, including pronouns for example.
> 
> I don't see how anyone benefits from an overly complicated solution for a problem we've solved a long time ago: Just adding your pronouns in the metadata

what i'd suggest is that, over the profile customization options, there could be some examples like

`You may add a location, link, another profile, pronouns, etc...`

but then i realized that this is already a thing!

![image](https://github.com/mastodon/mastodon/assets/65043391/76d63e43-cb29-47d9-beee-9e8284f56883)
(this is localized sorry, maybe someone can contribute with an english screenshot?)

**Comment by ThatOneCalculator** [comment]
> I don't see how anyone benefits from an overly complicated solution for a problem we've solved a long time ago: Just adding your pronouns in the metadata

@smiba my big push against that is being able to see someone's pronouns when replying to them without having to go to their profile

**Comment by jepyang** [comment]
@RealJosephKnapp
I’m not seeing how any of those criticisms are relevant. As far as I can tell in this point in the discussion:
- nobody is suggesting the creation of a new and fundamentally incompatible implementation of ActivityPub. All of this is possible within current activitypub specs, it’s not even a particularly significant departure from the existing mastodon software.
- Nobody is suggesting anything specific to neopronouns whatsoever. The goal seems to be to make the software flexible enough to accommodate them, not to lock down the software such that people who are disinclined are somehow forced into using them.

It…kinda seems like you saw neopronouns mentioned and just freaked for no reason.

@smiba 
My favored solution is to use the existing metadata fields and hardcode one of them as “pronouns” instead of making it user-assignable. It could be visually separate from the other metadata fields in the settings screen, ignored if left blank, and different clients could render it however they prefer, knowing that if something is there, it should be labeled “pronouns.” The profiles of people who don’t use the field are left unchanged.

**Comment by ThatOneCalculator** [comment]
> what's your opinion on viewing the profile fields on hover? (can you already do that anywhere else?)

That's already a thing in Firefish/Misskey.

**Comment by ThatOneCalculator** [comment]
@jepyang agreed. Their comment shouldn't be considered in this discussion as productive and marked as off-topic.

**Comment by derspyy** [comment]
> > what's your opinion on viewing the profile fields on hover? (can you already do that anywhere else?)
> 
> That's already a thing in Firefish/Misskey.

if implemented into the official client it would at least remove some stress from this issue, right? just as a side thought, though.

ignoring the previous guy, my main concern is how "hardcoding" the field will work through different localizations.
i'm not knowledgeable enough about activitypub but i believe that the field would have to be sent as a string like "Pronouns" to various services, and having it in english by default doesn't sound very good. it could default to your interface language?

something else that bothers me about it is this:

@12345axc 
> I think we should take account that some language doesn't have gendered pronoun :))
> Like Indonesian who only have "dia" for every gender.

really want to emphasize i'm raising doubts as a way to push this discussion forwards, not to be petty about pronouns.

**Comment by ThatOneCalculator** [comment]
> if implemented into the official client it would at least remove some stress from this issue, right? just as a side thought, though.

Misskey nor Firefish are Mastodon clients. They're completely different software that uses the ActivityPub protocol.

**Comment by derspyy** [comment]
> > if implemented into the official client it would at least remove some stress from this issue, right? just as a side thought, though.
> 
> Misskey nor Firefish are Mastodon clients. They're completely different software that uses the ActivityPub protocol.

i'm sorry for the confusion. i meant to ask your opinion on mastodon doing the same for quick access to pronouns.

**Comment by ThatOneCalculator** [comment]
Ah. Yeah, that'd be good. The first couple metadata are usually important anyway, even if they're not pronouns.

**Comment by derspyy** [comment]
#14918 : p

**Comment by nachtjasmin** [comment]
For those of you who are using the browser, there's a [browser extension called ProToots](https://github.com/ItsVipra/ProToots/) that might serve as a bridge until the feature is implemented natively.

**Comment by shleeable** [comment]
forgejo has added it https://codeberg.org/forgejo/forgejo/pulls/1518

**Comment by deanveloper** [comment]
Something that @sk22 does for the Megalodon Android app is display the "pronouns" field of a user next to their name. Very nice feature, I wish all Mastodon clients had this.

Here's a (largely censored) image of the pronouns fields:
![Screenshot of the "Home" timeline of the Megalodon App with all usernames, icons, and post content censored. Has large arrows pointing at all of the users' pronouns](https://github.com/mastodon/mastodon/assets/3196327/140378bc-9b89-43f1-b878-baf7b850f1f8)

**Comment by derspyy** [comment]
> Something that @sk22 does for the Megalodon Android app is display the "pronouns" field of a user next to their name. Very nice feature, I wish all Mastodon clients had this.
> 
> 
> 
> Here's a (largely censored) image of the pronouns fields:
> 
> ![Screenshot of the "Home" timeline of the Megalodon App with all usernames, icons, and post content censored. Has large arrows pointing at all of the users' pronouns](https://github.com/mastodon/mastodon/assets/3196327/140378bc-9b89-43f1-b878-baf7b850f1f8)
> 
> 

i think that any solution in this format must take into account localization to be fully accessible, which i think is the main challenge here.

**Comment by smiba** [comment]
> i think that any solution in this format must take into account localization to be fully accessible, which i think is the main challenge here.


This would have to happen either way, because if someone writes their pronouns in their native language I wouldn't know how to use them in English. If someone would write "mes pronoms sont il/lui" (French), I'd have no clue what these might be in English

**Comment by derspyy** [comment]
> > i think that any solution in this format must take into account localization to be fully accessible, which i think is the main challenge here.
> 
> 
> 
> 
> 
> This would happen either way, because if someone writes their pronouns in their native language I wouldn't know how to use them in English. If someone would write "mes pronoms sont il/lui" (French), I'd have no clue what these might be in English, as I don't speak French.

yeah, but there are people who know more than one language. we'd want the pronoun field to support both english and french.

**Comment by sk22** [comment]
regarding internationalization: the way megalodon does it is that it checks for a profile field that matches whatever "pronouns" means in the user's display language - with the english "pronouns" as a fallback

**Comment by sk22** [comment]
and tbh, the main challenge wasn't the internationalization (i mean maybe my approach was too simple anyway), but trying to extract the actual pronouns from the field; for example, stripping away any url from values like "they/them - https://pronoun.is/they/them"

**Comment by derspyy** [comment]
> and tbh, the main challenge wasn't the internationalization (i mean maybe my approach was too simple anyway), but trying to extract the actual pronouns from the field; for example, stripping away any url from values like "they/them - https://pronoun.is/they/them"

true... i think the ways people might have their pronouns setup are a bit too dynamic, considering languages, links, formats, etc. which is why i proposed a way for us to just have a quick glance at all of the user's fields. (though this isn't super ideal in mobile UX, i'm guessing)

**Comment by Wuzzy2** [comment]
@sk22: Frankly, this sounds like a horrible hacky solution. Implementations that try to be smarter than the user or read their mind usually fail in some weird edge case.
The real solution is not at the client side with random guessing, but in Mastodon itself. Everything else is just a hack.

**Comment by deanveloper** [comment]
@Wuzzy2 Its a solution that was born out of restrictions -- the Megalodon client has no control over how other mastodon users set pronouns, but it took common patterns and tried to adapt them into a way to display pronouns in the best (but not perfect) way to the user.

Obviously a more robust solution is needed, but the Megalodon solution is a jumping pad that we can use. For instance, what if the user metadata table had a drop-down for the key, and the drop-down included things like "pronouns" "website" and "other"? (a solution similar to the one on the GitHub profile page). Then, if pronouns is selected, another dropdown (multiselect?) can be displayed, offering a standardized way to add pronouns to a user profile, allowing more computer-readable pronouns so they can be displayed in other places

**Comment by RiedleroD** [comment]
I would like to note that there is an AP extension for this: https://ns.pancakes.gay/as/#pronouns

I'm told that IceShrimp.NET and wafrn already support this, while it's currently [being discussed](https://akkoma.dev/AkkomaGang/akkoma-fe/issues/363) for akkoma.

Would be really nice if we could have pronouns in fedi before long. If mastodon implements this, many others are sure to follow.

(sidenote but how pixelfed implemented this is pretty terrible and unnecessarily restrictive… please just put a plaintext field instead, even if you don't wanna impl the previously mentioned AP ex)

**Comment by smiba** [comment]
> I personally think the profile metadata fields are fine for this and make sense. Adding a separate pronoun feature would add clutter and would personally not feel much different then what we have right now?

Coming back to my comment from 2,5 years ago, I actually agree with this proposal nowadays. 
I think this is something that should be considered
